src/extension.ts

```diff
@@ -262,35 +264,61 @@ export class Extension implements RunHooks {
       this._treeItemObserver,
       registerTerminalLinkProvider(this._vscode),
     ];
-    const fileSystemWatchers = [
-      // Glob parser does not supported nested group, hence multiple watchers.
-      this._vscode.workspace.createFileSystemWatcher('**/*playwright*.config.{ts,js,mts,mjs}'),
-      this._vscode.workspace.createFileSystemWatcher('**/*.env*'),
-    ];
-    this._disposables.push(...fileSystemWatchers);
 
-    const rebuildModelForConfig = (uri: vscodeTypes.Uri) => {
-      // TODO: parse .gitignore
-      if (uriToPath(uri).includes('node_modules'))
-        return;
-      if (!this._isUnderTest && uriToPath(uri).includes('test-results'))
-        return;
-      void this._rebuildModels(false);
-    };
+    const configObserver = new WorkspaceObserver(this._vscode, async change => {
+      const hasRelevantFile = [...change.created, ...change.changed, ...change.deleted].some(path => {
+        // TODO: parse .gitignore
+        if (path.includes('node_modules'))
+          return false;
+        if (!this._isUnderTest && path.includes('test-results'))
+          return false;
+        return true;
+      });
+      if (hasRelevantFile)
+        void this._rebuildModels();
+    });
+    configObserver.setPatterns(new Set([
+      '**/*playwright*.config.{ts,js,mts,mjs}',
+      '**/*.env*',
+    ]));
+    this._disposables.push(configObserver);
+    await this._rebuildModelsImmediately(false);
 
-    await this._rebuildModels(false);
-    fileSystemWatchers.map(w => w.onDidChange(rebuildModelForConfig));
-    fileSystemWatchers.map(w => w.onDidCreate(rebuildModelForConfig));
-    fileSystemWatchers.map(w => w.onDidDelete(rebuildModelForConfig));
     this._context.subscriptions.push(this);
   }
 
-  private async _rebuildModels(userGesture: boolean): Promise<vscodeTypes.Uri[]> {
-    this._commandQueue = Promise.resolve();
+  private async _rebuildModels() {
+    if (this._modelRebuild) {
+      this._modelRebuild.needsAnother = true;
+      return;
+    }
+
+    await this._rebuildModelsImmediately(false);
+  }
+
+  private async _rebuildModelsImmediately(userGesture: boolean) {
+    this._modelRebuild?.token.cancel();
+    await this._modelRebuild?.result;
+
+    const cancel = new this._vscode.CancellationTokenSource();
+    const rebuild = { result: this._innerRebuildModels(userGesture, cancel.token), token: cancel, needsAnother: false };
+    this._modelRebuild = rebuild;
+    await this._modelRebuild.result.finally(() => {
+      if (this._modelRebuild === rebuild)
```

Comment:
The fact that we compare these two suggests we expect some race? These two should always match!